.github/workflows/serverless-service.yml

```diff
@@ -3,80 +3,113 @@
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+      - name: Install dependencies
+        run: make dev
+      - name: pre commit
+        run: make pre-commit
+      - name: Formatting and Linting
+        run: make lint
+      - name: Complexity scan
+        run: make complex
+  tests:
+    needs: quality_standards
     runs-on: ubuntu-latest
     environment: dev
+    permissions:
+      id-token: write  # required for requesting the JWT (GitHub OIDC)
     steps:
-      - run: echo "🎉 The job was automatically triggered by a ${{ github.event_name }} event."
-      - run: echo "🐧 This job is now running on a ${{  runner.os }} server hosted by GitHub!"
-      - run: echo "🔎 The name of your branch is ${{ github.ref }} and your repository is ${{ github.repository }}."
       - name: Check out repository code
         uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
       - name: Set up Python 3.11
         uses: actions/setup-python@61a6322f88396a6271a6ee3565807d608ecaddd1 # v4.7.0
         with:
-          python-version: "3.11"
+          python-version: ${{ env.PYTHON_VERSION }}
       - name: Set up Node
         uses: actions/setup-node@bea5baf987ba7aa777a8a0b4ace377a21c45c381 # v3.8.0
         with:
-          node-version: "18"
+          node-version: ${{ env.NODE_VERSION }}
       - name: Install dependencies
-        run: |
-          pip install --upgrade pip pre-commit poetry
-          npm install -g aws-cdk
-          make deps
-          pip install -r dev_requirements.txt
-          pip install -r lambda_requirements.txt
-      - name: configure aws credentials
+        run: make dev
+      - name: Unit tests
+        run: make unit
+      - name: Infrastructure tests
+        run: make infra-tests
+      - name: Configure AWS credentials
         uses: aws-actions/configure-aws-credentials@5fd3084fc36e372ff1fff382a39b10d03659f355 # v2.2.0
         with:
           role-to-assume: ${{ secrets.AWS_ROLE }}
-          role-session-name: samplerolesession
-          aws-region: us-east-1
-      - name: pre commit
-        run: |
-          make pre-commit
-      - name: Lint with flake8 and mypy
-        run: |
-          make lint
-      - name: file format
-        run: |
-          yapf -d -vv --style=./.style --exclude=.venv --exclude=.build --exclude=cdk.out --exclude=.git  -r .
-      - name: Complexity scan
-        run: |
-          make complex
-      - run: echo "🖥️ The workflow is now ready to test your code on the runner."
-      - name: Unit tests
-        run: |
-          make unit
-      - name: Infrastructure tests
-        run: |
-          make infra-tests
+          role-session-name: ${{ env.SESSION_NAME }}
+          aws-region: ${{ env.AWS_REGION }}
+        env:
+          SESSION_NAME: "${{github.repository}}-at-${{github.sha}}"
       - name: Deploy to AWS
         run: make deploy
+      # NOTE: these run unit and integration tests
```

Comment:
i see that you separated build and test, which is cool , it just means that you setup the env twice which makes the pipelines longer, that's why i didnt seperate them, but it makes sense

Reply:
I've just added caching to make this minimal in all 3 job steps that were missing. Besides perf, separation was also to keep jobs with minimal privileges - I'd go even further and separate integ/e2e tests too and block sockets on unit/infra tests, but too many changes for now (we've got bigger fishes to fry).

```yaml
      - name: Set up Python
        uses: actions/setup-python@61a6322f88396a6271a6ee3565807d608ecaddd1 # v4.7.0
        with:
          python-version: ${{ env.PYTHON_VERSION }}
          cache: "poetry"

      - name: Set up Node
        uses: actions/setup-node@bea5baf987ba7aa777a8a0b4ace377a21c45c381 # v3.8.0
        with:
          node-version: ${{ env.NODE_VERSION }}
          cache: "npm"
```